doc/security/cryptolib/cryptolib_api.md

```diff
@@ -1,149 +1,123 @@
+The library is written in C and uses OpenTitan's hardware blocks for accelerated cryptography.
+It generally attempts to minimize code size and protect against side-channel and fault-injection attacks, including by physically present attackers.
+
+**Note: at the time of writing, the crypto library is still under development, and not all algorithms described in this page are fully implemented and tested.**
+
+## Supported Algorithms
+
+The following is a quick reference for algorithms and modes supported by the OT cryptolib.
+For more details, see later sections (links in the "category" column).
+
+| Category        | Supported schemes         |
+|-----------------|---------------------------|
+| [**AES**](#aes) | AES-{ECB,CBC,CFB,OFB,CTR}<br>AES-KWP<br>AES-GCM |
+| [**Hash functions**](#hash-functions) | SHA2-{256,384,512}<br>SHA3-{224,256,384,512}<br>SHAKE{128,256} (XOF)<br>cSHAKE{128,256} (XOF) |
+| [**Message authentication**](#message-authentication) | HMAC-SHA256<br>KMAC{128,256} |
+| [**RSA**](#rsa) | RSA-{2048,3072,4096} |
+| [**Elliptic curve cryptography**](#elliptic-curve-cryptography) | ECDSA-{P256,P384}<br>ECDH-{P256,P384}<br>Ed25519<br>X25519 |
+| [**Deterministic random bit generation**](#deterministic-random-bit-generation) | AES-CTR-DRBG |
+| [**Key derivation**](#key-derivation) | HMAC-KDF-CTR<br>KMAC-KDF-CTR |
+
+
+## Asynchronous operations
+
+For some functions, OpenTitan's cryptolib supports asynchronous calls.
+All operations which take longer than 10ms should have an asychronous interface.
+This is helpful for compatibility with TockOS, which has a low latency return call programming model.
+
+The OpenTitan cryptolib does not implement any thread management.
+Instead, it treats the OTBN coprocessor as a "separate thread" to achieve non-blocking operation with virtually zero overhead.
+OTBN sends an interrupt when processing is complete.
+
+All asynchronous operations have two functions:
+
+- **\<algorithm\>\_async\_start**
+    - Takes input arguments. Checks that OTBN is available and idle. If
+      so: does any necessary synchronous preprocessing, initializes
+      OTBN, and starts the OTBN routine.
+- **\<algorithm\>\_async\_finalize**
+    - Takes caller-allocated output buffers. Blocks until OTBN is done
+      processing if needed, then checks whether it had errors. If not, does any
+      necessary postprocessing and writes results to the buffers.
+
+The caller should call the `start` function, wait for the interrupt, and then call `finalize`.
+
+A few noteworthy aspects of this setup:
+- While an asynchronous operation is running, OTBN will be unavailable and attempts to use it will return errors.
+- Only one asynchronous operation may be in progress at any given time.
+- The caller is responsible for properly managing asynchronous calls, including ensuring that the entity receiving the `finalize` results is the same as the one who called `start`.
```

Comment:
Thanks for these changes @jadephilipoom - everything is great in terms of the refocusing of the documentation and I think it makes this much more user-friendly.

One small comment about this section here - do we want to expand on this and say what happens (or how we might mitigate against) an incorrect `finalize` call here?  (Either maliciously placed or inadvertent/bug-induced?)

Reply:
Some of that depends on the specifics of the OTBN program and the async algorithm that was called. There are some library-level mitigations we could do -- some ideas on https://github.com/lowRISC/opentitan/issues/19173 but none of these are yet implemented. Right now we don't really mitigate against this case and, if an attacker or another application calls `finalize()` before the original caller and no one runs a new OTBN operation after that, the original caller will silently get random data when they call `finalize()`.